File: Data Structures & Algorithms/valid-binary-search-tree/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public boolean isValidBST(TreeNode root) {
        long min = Long.MIN_VALUE;
        long max = Long.MAX_VALUE;

        return dfs(root, min, max);
    }


    public boolean dfs(TreeNode root, long min, long max){

        if(root == null) return true;

        if(root.val <= min || root.val >= max) return false;

        return dfs(root.left, min, root.val) && dfs(root.right, root.val, max);
    }
}

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-0.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        frequency = {}
        maxlength = 0
        maxfrequency = 0
        left = 0

        for right, c in enumerate(s):
            frequency[c] = frequency.get(c, 0) + 1

            maxfrequency = max(maxfrequency, frequency[c])

            while (right - left + 1) - maxfrequency > k: 
                frequency[s[left]] -= 1
                left += 1
            
            maxlength = max(maxlength, (right  - left + 1))
        
        return maxlength

